src/drv_vrtek/vrtek.c

```diff
@@ -0,0 +1,552 @@
+        }
+    }
+    else {
+        priv->sku = VRTEK_SKU_UNKNOWN_UNKNOWN;
+        LOGE("Unknown HMD model: %s\n", priv->model_str);
+        return;
+
+    }
+
+    LOGI("HMD model is %s (%s)", vrtek_get_model_string(priv->sku),
+         priv->model_str);
+}
+
+static void vrtek_set_imu_state(vrtek_priv* priv, bool state)
+{
+    uint8_t response_data[RESPONSE_DATA_SIZE];
+    uint8_t response_data_len;
+
+    uint8_t imu_state = state;  /* 1 = enabled, 0 = disabled */
+
+    vrtek_send_command_and_receive_response(priv, VRTEK_COMMAND_CONFIGURE_HMD,
+                                            VRTEK_CONFIG_ARG_HMD_IMU,
+                                            &imu_state, sizeof(uint8_t),
+                                            response_data, &response_data_len);
+    assert(response_data_len == 1);
+    assert(response_data[0] == 0);
+
+    LOGI("HMD IMU %s", imu_state ? "enabled" : "disabled");
+
+    ohmd_sleep(0.1);
+}
+
+static void vrtek_read_config(vrtek_priv* priv)
+{
+    uint8_t response_data[RESPONSE_DATA_SIZE];
+    uint8_t response_data_len;
+
+    vrtek_send_command_and_receive_response(priv, VRTEK_COMMAND_QUERY_HMD,
+                                            VRTEK_QUERY_ARG_HMD_CONFIG,
+                                            NULL, 0,
+                                            response_data, &response_data_len);
+    assert(response_data_len == 26);
+
+    /* display type is the second byte of the response */
+    uint8_t display_type = response_data[1];
+    vrtek_update_display_properties(priv, display_type);
+    vrtek_update_hmd_software_version(priv, response_data, response_data_len);
+}
+
+static void vrtek_read_model(vrtek_priv* priv)
+{
+    uint8_t response_data[RESPONSE_DATA_SIZE];
+    uint8_t response_data_len;
+
+    vrtek_send_command_and_receive_response(priv, VRTEK_COMMAND_QUERY_HMD,
+                                            VRTEK_QUERY_ARG_HMD_MODEL,
+                                            NULL, 0,
+                                            response_data, &response_data_len);
+    assert(response_data_len == 56);
+
+    const char* model = (const char*) response_data;
+    vrtek_update_hmd_model(priv, model, response_data_len);
+}
+
+static void update_device(ohmd_device* device)
+{
+    int size = 0;
+    uint8_t buf[REPORT_BUFFER_SIZE];
+    vrtek_priv* priv = vrtek_priv_get(device);
+
+    while ((size = hid_read(priv->hid_handle, buf, REPORT_BUFFER_SIZE)) > 0) {
+        if (buf[0] == VRTEK_REPORT_SENSOR) {
+            vrtek_decode_hmd_data_packet(buf, size, &priv->hmd_data);
+        } else {
+            LOGE("unknown message type: %u", buf[0]);
+        }
+    }
+
+    if (size < 0) {
+        LOGE("error reading from device");
+    }
+}
+
+static int getf(ohmd_device* device, ohmd_float_value type, float* out)
+{
+    vrtek_priv* priv = vrtek_priv_get(device);
+
+    switch (type) {
+    case OHMD_ROTATION_QUAT:
+        *(quatf*)out = *(quatf*)&priv->hmd_data.quaternion;
```

Comment:
You are using direct quaternion here, while accel and gyro are available from the HMD.
We prefer to use our own fusion, since this allows us to do updates and tweaks where the pre-fusion by the firmware might be inadequate. Please look at other drivers where fusion is used and implement this.

Reply:
That makes total sense - ensure we have control rather than relying on an unknown black box. I'll work on this next. I may come find you folks on IRC if I have questions :)

Thanks!